### PR TITLE
Backport relevant master branch code to 4.10.x

### PR DESCRIPTION
**Pull request by Iristyle**
The following is a list of actions performed based on PRs merged to master which should land in the 4.10.x release - more info at https://gist.github.com/Iristyle/8a3236d04e246a7a1159c188fb4ca133

#### A merge-up to master results in only 2 conflicts:

* `locales/puppet.pot` based on automation generating translations
* `lib/puppet/pops/types/ruby_generator.rb` due to PUP-7212 / PUP-7222 / PUP-7227 code landing

#### Features / refactors / other code changes that could plausibly be Puppet 4

* [PR 5472 - (PUP-4283) Update Puppet Module Tool to use SemanticPuppet::Version](http://github.com/puppetlabs/puppet/pull/5472) - [PUP-4283](https://tickets.puppetlabs.com/browse/PUP-4283)
    * git cherry-pick 4d297235bb666dd2770955f1740af0bc91eb4b8d
        * resolved conflicts

#### Features / refactors / other code changes that could plausibly be Puppet 4

* [PR 5505 - (PUP-5027) Prevent that metadata.json is parsed multiple times](http://github.com/puppetlabs/puppet/pull/5505) - [PUP-5027](https://tickets.puppetlabs.com/browse/PUP-5027)
    * git cherry-pick f2e8e6649d392d5d3b7b8e79efbf15815adcdcc8
        * resolved conflicts
* [PR 5504 - (PUP-7100) Add Hiera version 5 lookup_key function for eyaml](http://github.com/puppetlabs/puppet/pull/5504) - [PUP-7100](https://tickets.puppetlabs.com/browse/PUP-7100)
    * this code was already brought into stable via another backport - skipping the 5 commits:
    * b42ef28cfec54a6072a489b5a631624cbd42d92f
    * 26ca009b5aeca8d4b8c5e3057ceb79830b467ba5
    * 078273acbe99cff80d35d805e6bd168ff189f607
    * a7790be2c3e16999bf6bdb3bd29e3cc627f95ae4
    * 66457e73e076cf19589f90d17626f95457e22a78

#### Bugfixes that should probably be Puppet 4

* [PR 5557 - (PUP-6006) Change phrasing of multiple errors final error](http://github.com/puppetlabs/puppet/pull/5557) - [PUP-6006](https://tickets.puppetlabs.com/browse/PUP-6006)
    * git cherry-pick efcdf580b1013541da3f92104e0ac3ebac52f901

#### Glisan Agent UTF-8

* [PR 5522 - (PUP-7122) Add wrapper for gettext:update_pot rake task](http://github.com/puppetlabs/puppet/pull/5522) - [PUP-7122](https://tickets.puppetlabs.com/browse/PUP-7122)
    * git cherry-pick d32d3e26ea624a4d1ac7f0bd83b0fcf5e242814d
    * note: only took one commit as i18n.rake no longer exists

#### Bugfixes that should probably be Puppet 4

* [PR 5615 - (PUP-7168) Add `#==` alias for `#eql?` method of Pcore Objects](http://github.com/puppetlabs/puppet/pull/5615) - [PUP-7168](https://tickets.puppetlabs.com/browse/PUP-7168)
    * git cherry-pick 36d98197a60d388cd47055ed3e413e77319dc37b

#### Features / refactors / other code changes that could plausibly be Puppet 4

* [PR 5502 - (PUP-6494) Support sensitive commands in Exec resource type.](http://github.com/puppetlabs/puppet/pull/5502) - [PUP-6494](https://tickets.puppetlabs.com/browse/PUP-6494)
    * git cherry-pick e1dd130100316f7fcaf7f2dc095f6b50c98cc6be

#### Bugfixes that should probably be Puppet 4

* [PR 5546 - (PUP-7129) Use systemd service provider for Debian 9 (Strech)](http://github.com/puppetlabs/puppet/pull/5546) - [PUP-7129](https://tickets.puppetlabs.com/browse/PUP-7129)
    * git cherry-pick e20c4bd2a5a38fe34ef052be37be254db2a1d965
    * git cherry-pick 5bd1a659f7e8b3d57028f964d07f37575e8f9f90

#### Glisan Agent UTF-8    

* [PR 5594 - (PUP-7152) AIX encrypted user passwords are ASCII](http://github.com/puppetlabs/puppet/pull/5594) - [PUP-7152](https://tickets.puppetlabs.com/browse/PUP-7152)
    * git cherry-pick d6aaca5877b20f0772aaaba7ceefa19f93f72cdc
* [PR 5593 - (PUP-7151) Read selinux module version as BINARY](http://github.com/puppetlabs/puppet/pull/5593) - [PUP-7151](https://tickets.puppetlabs.com/browse/PUP-7151)
    * git cherry-pick 80af250667b05d37c65dcd2cd50973a25c1be325

#### Bugfixes that should probably be Puppet 4

* [PR 5632 - (PUP-7191) consider non-existent services on AIX to be stopped/disabled rather than failing](http://github.com/puppetlabs/puppet/pull/5632) - [PUP-7191](https://tickets.puppetlabs.com/browse/PUP-7191)
    * git cherry-pick 68aea974d7144bb6e6830ccfe92774f6315361fe
    * git cherry-pick 55a67e6b1b6d7246cba921b6d8514fb381dab83c
    * git cherry-pick b63c540cc205eb1e897db9d33da8111d859e2847
    * git cherry-pick 55a67e6b1b6d7246cba921b6d8514fb381dab83c

#### Glisan Agent UTF-8

* [PR 5384 - (PUP-7134) Use Puppet::FileSystem for file access / explicitly specify encodings for SSL files ](http://github.com/puppetlabs/puppet/pull/5384) - [PUP-7134](https://tickets.puppetlabs.com/browse/PUP-7134)
    * git cherry-pick dfebcf700cd17ee7f2bc33374d55fff9568aa05e
    * git cherry-pick a8406be58b8bef6cf8cce843cdeff692fe916125
    * git cherry-pick f0dab056df37adcb1fd014df1a13620aaac52722
    * git cherry-pick 4ad7717fbf0453466a62af1b49a776afd93a5bb9

#### Bugfixes that should probably be Puppet 4

* [PR 5639 - (PUP-7214) PUP-6663 causes a regression of PUP-535](http://github.com/puppetlabs/puppet/pull/5639) - [PUP-7214](https://tickets.puppetlabs.com/browse/PUP-7214)
  * Note - this was backported to Puppet 4 already
* [PR 5650 - (PUP-7102) Fail with error on access problems with environment directory](http://github.com/puppetlabs/puppet/pull/5650) - [PUP-7102](https://tickets.puppetlabs.com/browse/PUP-7102)
    * GIVEN NEXT PR REVERTED, DID NOT PERFORM - git cherry-pick e4a4e09323436c5e44bc3b3154f42cd020c05a92
    * GIVEN NEXT PR REVERTED, DID NOT PERFORM - git cherry-pick ae58568128348a4a425444b98cec78e9420f13a9
* [PR 5667 - Revert "(PUP-7102) Fail with error on access problems with environment directory"](https://github.com/puppetlabs/puppet/pull/5667)
    * GIVEN THIS REVERTED PRIOR PR, DID NOT PERFORM - git cherry-pick b1d12540ae06943e3a793d5136ee1fe57b90684d

#### Glisan Agent UTF-8

* [PR 5633 - (pup-7021) establish central string encoding conversion](http://github.com/puppetlabs/puppet/pull/5633) - [pup-7021](https://tickets.puppetlabs.com/browse/pup-7021)
    * git cherry-pick df46662011d00f95345a9e846115042de1249166
    * git cherry-pick 346036876996ff1a6d7c169a818c7453c0cc47a6

#### MISC
* [(maint) fix minor spec regex warning](https://github.com/puppetlabs/puppet/pull/5661)
    * git cherry-pick 56cda69cbed18b3ce5fc2037efe338256165f67f

#### Removals / changes that should wait for Puppet 5

* [PR 4999 - (PUP-6372) Remove unused Puppet::Transaction.failed? method](http://github.com/puppetlabs/puppet/pull/4999) - [PUP-6372](https://tickets.puppetlabs.com/browse/PUP-6372)
* [PR 5479 - (PUP-5434) Remove deprecated experimental lookup implementation](http://github.com/puppetlabs/puppet/pull/5479) - [PUP-5434](https://tickets.puppetlabs.com/browse/PUP-5434)
* [PR 5491 - (PUP-6481) Remove external_facts feature](http://github.com/puppetlabs/puppet/pull/5491) - [PUP-6481](https://tickets.puppetlabs.com/browse/PUP-6481)
* [PR 5492 - (PUP-6219) Remove deprecated req_bits setting](http://github.com/puppetlabs/puppet/pull/5492) - [PUP-6219](https://tickets.puppetlabs.com/browse/PUP-6219)
* [PR 5493 - (PUP-3670) Remove Facts#strip_internal method](http://github.com/puppetlabs/puppet/pull/5493) - [PUP-3670](https://tickets.puppetlabs.com/browse/PUP-3670)
* [PR 5494 - (PUP-4933) Remove deprecated cfacter setting](http://github.com/puppetlabs/puppet/pull/5494) - [PUP-4933](https://tickets.puppetlabs.com/browse/PUP-4933)
* [PR 5642 - (PUP-5973) Remove symbol string monkey patch](http://github.com/puppetlabs/puppet/pull/5642) - [PUP-5973](https://tickets.puppetlabs.com/browse/PUP-5973)
* [PR 5616 - Revert "Merge pull request #5276 from DavidS/notify-everywhere (PUP-6783)"](http://github.com/puppetlabs/puppet/pull/5616)
* [PR 5619 - (PUP-7142) Autoload Pcore types from TypeSet.](http://github.com/puppetlabs/puppet/pull/5619) - [PUP-7142](https://tickets.puppetlabs.com/browse/PUP-7142)
* [PR 5626 - (maint) Remove redundant code from AST Model Factory](http://github.com/puppetlabs/puppet/pull/5626)
* [PR 5641 - (PUP-7212) Add a Pcore Annotation type and an annotate function](http://github.com/puppetlabs/puppet/pull/5641) - [PUP-7212](https://tickets.puppetlabs.com/browse/PUP-7212)
* [PR 5645 - (PUP-7222) Add RubyMethod annotation and let RubyGenerator act on it](http://github.com/puppetlabs/puppet/pull/5645) - [PUP-7222](https://tickets.puppetlabs.com/browse/PUP-7222)
* [PR 5646 - (PUP-7227) Add "reference" as a valid Pcore Object attribute kind](http://github.com/puppetlabs/puppet/pull/5646) - [PUP-7227](https://tickets.puppetlabs.com/browse/PUP-7227)

#### Breaking changes that must wait for Puppet 5

* [PR 4584 - (PUP-5609) Protect shared state with filesystem lock](http://github.com/puppetlabs/puppet/pull/4584) - [PUP-5609](https://tickets.puppetlabs.com/browse/PUP-5609)
* [PR 5480 - (PUP-6112) Make trusted_server_facts always be true](http://github.com/puppetlabs/puppet/pull/5480) - [PUP-6112](https://tickets.puppetlabs.com/browse/PUP-6112)
* [PR 5610 - (PUP-7173) server_set facts rb failing with undefined local variable or method `master opts](http://github.com/puppetlabs/puppet/pull/5610) - [PUP-7173](https://tickets.puppetlabs.com/browse/PUP-7173)
* [PR 5510 - (PUP-4947) Make app management keywords standard](http://github.com/puppetlabs/puppet/pull/5510) - [PUP-4947](https://tickets.puppetlabs.com/browse/PUP-4947)
* [PR 5603 - (maint) update version.rb for 5.0](http://github.com/puppetlabs/puppet/pull/5603)
* [PR 5617 - (PUP-7175) lookup.rb failing automatic parameter lookup with module data binding](http://github.com/puppetlabs/puppet/pull/5617) - [PUP-7175](https://tickets.puppetlabs.com/browse/PUP-7175)
* [PR 5556 - (PUP-3170) Cap integer size](http://github.com/puppetlabs/puppet/pull/5556) - [PUP-3170](https://tickets.puppetlabs.com/browse/PUP-3170)
    * git cherry-pick 112e3b3ac03394258ac0b47d47b7faff5638c2c4
    * git cherry-pick 956f7556870fe90e54d7a6fae17b410ed509422c
    * git cherry-pick 0a5c85a109889fe70535dc261e27245e82a76fc5
    * git cherry-pick 189a6f8b8f21919353e6f650eea2e8fc59e623a2
    * git cherry-pick 67f22dfd0262009dd740b7d148e89c2050add3df
    * git cherry-pick 33252a1c9a938cb19366992e9b9f306950a314ff
    * git cherry-pick e2ba08435e7487e641a437b0ed9f111ae92c4e7e
    * git cherry-pick 5094ac2a0f0d6d6e1bcc4c70a3f89b47fac1e783
* [PR 5511 - (PUP-5954) Make class merge an error](http://github.com/puppetlabs/puppet/pull/5511) - [PUP-5954](https://tickets.puppetlabs.com/browse/PUP-5954)
    * git cherry-pick 75a81f6cb9e81b4b3094c1e6324c462da6f61f43
    * git cherry-pick 0a36704e2f7a6ca09ce84b7ea471ab45031434a8
    * git cherry-pick e9356dd36f47ded66e82d5174c2f035fc5c0ddcb
* [PR 5477 - (PUP-6930) Make DUPLICATE_DEFAULT an error](http://github.com/puppetlabs/puppet/pull/5477) - [PUP-6930](https://tickets.puppetlabs.com/browse/PUP-6930)
    * git cherry-pick c2d14d8359ac51d65e881f2e6b33b67a0ed4ccd4
    * git cherry-pick 485518c0353525a1dab3639ffe3e70211551af25

### CODE REVIEW

**Comment by Iristyle** [comment]
### git log --oneline --decorate origin/stable..4.10.x

```
9d67eff (HEAD -> 4.10.x, Iristyle/4.10.x) (maint) fix minor spec regex warning
b7269fc (PUP-7021) i18n wrap string in character_encoding.rb
b22adb2 (pup-7021) establish central string encoding conversion
846920d (PUP-7102) Fail with error on access problems with environment directory
f15026c (PUP-7102) Make MemoryFile respond to #stat
7faa87a (PUP-7134) Use Puppet::FileSystem in ssl
6dae736 (PUP-7134) Puppet::SSL::Key explicit encodings
bedd54f (PUP-7134) Add csr_attributes UTF-8 tests
4650056 (PUP-7134) UTF-8 test user type authorized_keys
98eed8d (PUP-7191) update acceptance for non-existent AIX service
c0613a1 (maint) remove unreached line in aix service provider
efb261c (PUP-7191) consider non-existent services 'stopped' on AIX
56f75e4 (maint) whitespace fixes in aix service provider
5707e82 (PUP-7151) Read selinux module version as BINARY
fa13af4 (PUP-7152) AIX encrypted user passwords are ASCII
79f7560 (PUP-7129) Add support for stretch/sid and buster/sid
0c1748e (PUP-7129) Use systemd service provider for Debian 9 (Strech)
e416fb1 (PUP-6494) Support sensitive commands in Exec resource type.
e93b58b (PUP-7168) Add `#==` alias for `#eql?` method of Pcore Objects
e863de7 (maint) Ignore vendored code when searching for marked strings
f1c873a (PUP-3170) Add tests for arithmetic overflow
e92e499 (PUP-3170) Change Factory to deal with Integer instead of Fixnum
07ad330 (PUP-3170) Add test for cap of Integer literal values
fea0cb8 (PUP-3170) Add literal integer cap and cap on arithmetic
6b55722 (maint) Fix indentation
78eba5d (PUP-3170) Move definition of MIN/MAX_INTEGER to Pops namespace
3cf5cdf (PUP-3170) Add test for nested Integer out of range.
c282b56 (PUP-3170) Add cap to Integer and forbid BigDecimal values
6291125 (PUP-6006) Change phrasing of multiple errors final error
00c7f4c (maint) mark TypeCollection class as @api private
bf568cf (PUP-5954) Make attempt to merge class other than '' an error
849f276 (PUP-5954) Change how 'settings' class is set in type collection
e6ef082 (PUP-5027) Prevent that metadata.json is parsed multiple times
d8f9ab1 (PUP-6930) Modify test for logic that is unaffected by --strict
b5a1537 (PUP-6930) Make DUPLICATE_DEFAULT an error
033957f (PUP-4283) Update Puppet Module Tool to use SemanticPuppet::Version
26644d8 (origin/4.10.x) Updating POT file for d7d63f8c1c596b6ff56a57f244205bcf2d975a15
d7d63f8 Merge branch 'stable' into 4.10.x
1efd231 Updating POT file for b30d12114011153c524c8bc3ed485f1b9c39481b
b30d121 Merge pull request #5666 from thallgren/issue/pup-7204/hiera-mapped-paths
7f7b0f7 (PUP-7204) Add tests asserting that paths are included in explain output.
98702fc (PUP-7204) Add tests for when variables are shadowed by mapped paths
08d76b7 (PUP-7204) Add support for mapped paths to hiera.yaml
1795af8 Merge commit '29c05d63e69a3f4ebbdb8a5f9abfd736fff7183e' into 4.10.x
29c05d6 Updating POT file for 9d9360e17852fb905e778302315cdb7b86702073
65137c3 Updating POT file for 9d9360e17852fb905e778302315cdb7b86702073
9d9360e Merge pull request #5698 from Magisus/remove-rake-wrappers
b549bcc (maint) Remove gettext rake task wrappers
ad99698 Merge pull request #5695 from hlindberg/PUP-7306_allow-newline-in-regexp
7bf37e8 (PUP-7306) Accept newline in literal regular expressions
5d05e7c Merge branch 'stable' into 4.10.x
794515a Merge branch 'stable' into 4.10.x
c3d5f90 Merge remote-tracking branch 'upstream/stable' into 4.10.x
542e938 (packaging) Bump version to 4.10.0
79dcc5b Merge branch 'stable' into 4.10.x
4c2a795 Merge remote-tracking branch 'upstream/stable' into 4.10.x
ad61f80 Merge pull request #5680 from Magisus/PUP-7042/fix-specs
4871e4e Merge branch 'stable' into 4.10.x
36eaa777 Merge branch 'stable' into 4.10.x
6056b12 (PUP-7240) Fix bug when using lookup_options with 'global_only' lookup
8961e31 (PUP-7240) Prevent invocation inheritance when lookup adpater differs
9a11bfc (PUP-7042) Fix some problematic strings
7b05d6d (PUP-7042) Fix spec failures
```

File: lib/puppet/environments.rb [hunk 1]
```diff
@@ -213,9 +213,34 @@ def create_environment(name)
...
+        while current
+          begin
+            break if Puppet::FileSystem.stat(current).readable?
+          rescue Errno::EACCES
+          end
+          prev = current
+          current = Puppet::FileSystem.dir(prev)
+        end
+        raise Errno::EACCES, prev.to_s
+      end
```
Comment (joshcooper) [inline lib/puppet/environments.rb]: We reverted this commit in https://github.com/puppetlabs/puppet/pull/5667. I think just cherry-pick the revert?

File: lib/puppet/interface.rb [hunk 2]
```diff
@@ -154,10 +154,7 @@ def initialize(name, version, &block)
     end
 
     @name    = Puppet::Interface::FaceCollection.underscorize(name)
-
-    # SemVer is deprecated but in 4.x we must use it here (the attr_reader is public api). The
-    # extra boolean argument suppresses the deprecation warning.
-    @version = SemVer.new(version, true)
+    @version = SemanticPuppet::Version.parse(version)
```
Comment (joshcooper) [inline lib/puppet/interface.rb]: I'm not sure we want this backported to 4.10.x as it's not i18n related? There does seem to be a perf benefit, see https://tickets.puppetlabs.com/browse/PUP-5027. /cc @thallgren

File: lib/puppet/parser/compiler.rb [hunk 3]
```diff
@@ -854,7 +854,8 @@ def set_node_parameters
 
   def create_settings_scope
     settings_type = Puppet::Resource::Type.new :hostclass, "settings"
-    environment.known_resource_types.add(settings_type)
+    # use replace instead of add to avoid duplication check and illegal merge (PUP-5954)
+    environment.known_resource_types.replace_settings(settings_type)
```
Comment (joshcooper) [inline lib/puppet/parser/compiler.rb]: I don't think we want this commit d8d2c9be as it causes an error instead of a deprecation warning. See https://tickets.puppetlabs.com/browse/PUP-5954

File: lib/puppet/pops.rb [hunk 4]
```diff
@@ -10,10 +10,13 @@ module Puppet
...
   # @api public
   module Pops
-     EMPTY_HASH = {}.freeze
+    EMPTY_HASH = {}.freeze
     EMPTY_ARRAY = [].freeze
     EMPTY_STRING = ''.freeze
 
+    MAX_INTEGER =  0x7fffffffffffffff
+    MIN_INTEGER = -0x8000000000000000
+
```
Comment (joshcooper) [inline lib/puppet/pops.rb]: Not sure if we want 59640dd6d (https://tickets.puppetlabs.com/browse/PUP-3170) in 4.10.x /cc @hlindberg @thallgren It adds a check against overflow/underflow.

File: lib/puppet/pops/validation/validator_factory_4_0.rb [hunk 5]
```diff
@@ -28,7 +28,6 @@ def severity_producer
     p[Issues::FUTURE_RESERVED_WORD]          = :deprecation
 
     p[Issues::DUPLICATE_KEY]                 = Puppet[:strict] == :off ? :ignore : Puppet[:strict]
-    p[Issues::DUPLICATE_DEFAULT]             = Puppet[:strict] == :off ? :ignore : Puppet[:strict]
```
Comment (joshcooper) [inline lib/puppet/pops/validation/validator_factory_4_0.rb]: I think we don't want commit c2d14d8359ac51d65e881f2e6b33b67a0ed4ccd4 as it makes it hard error always in 4.10.x.

File: lib/puppet/environments.rb (hunk 1 above)
Comment (Iristyle) [inline lib/puppet/environments.rb]: thanks - removed

File: lib/puppet/parser/compiler.rb (hunk 3 above)
Comment (Iristyle) [inline lib/puppet/parser/compiler.rb]: PUP-5954 related code has been removed from this PR

File: lib/puppet/pops.rb (hunk 4 above)
Comment (Iristyle) [inline lib/puppet/pops.rb]: PUP-3170 related code has been removed from this PR per @hlindberg

File: lib/puppet/pops/validation/validator_factory_4_0.rb (hunk 5 above)
Comment (Iristyle) [inline lib/puppet/pops/validation/validator_factory_4_0.rb]: Removed PUP-6930 related code

File: lib/puppet/interface.rb (hunk 2 above)
Comment (joshcooper) [inline lib/puppet/interface.rb]: @Iristyle and I talked about this, and this doesn't appear to hurt anything. It switches PMT to using semantic puppet that is already vendored in puppet (so there's no integration issue for puppetserver) and there's a perf benefit, so 👍 to including in 4.10.x

**Comment by MosesMendoza** [comment]
👍

**Comment by joshcooper** [comment]
@hlindberg @thallgren can you do the merge up to master? I'm getting a conflict in  `lib/puppet/pops/types/ruby_generator.rb` related to https://github.com/puppetlabs/puppet/pull/5615.

**Comment by Iristyle** [comment]
Note that a later commit reverted the work in #5472  / [PUP-4283](https://tickets.puppetlabs.com/browse/PUP-4283)
